receiver/skywalkingreceiver/config.go

```diff
@@ -0,0 +1,90 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package skywalkingreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/skywalkingreceiver"
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configgrpc"
+	"go.opentelemetry.io/collector/config/confighttp"
+)
+
+const (
+	// The config field id to load the protocol map from
+	protocolsFieldName = "protocols"
+)
+
+// Protocols is the configuration for the supported protocols.
+type Protocols struct {
+	GRPC *configgrpc.GRPCServerSettings `mapstructure:"grpc"`
+	HTTP *confighttp.HTTPServerSettings `mapstructure:"http"`
+}
+
+// Config defines configuration for skywalking receiver.
+type Config struct {
+	config.ReceiverSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+	Protocols               `mapstructure:"protocols"`
+}
+
+var _ config.Receiver = (*Config)(nil)
+var _ config.Unmarshallable = (*Config)(nil)
+
+// Validate checks the receiver configuration is valid
+func (cfg *Config) Validate() error {
+	if cfg.GRPC == nil && cfg.HTTP == nil {
+		return fmt.Errorf("must specify at least one protocol when using the Skywalking receiver")
+	}
+
+	if cfg.GRPC != nil {
+		var err error
+		if _, err = extractPortFromEndpoint(cfg.GRPC.NetAddr.Endpoint); err != nil {
+			return fmt.Errorf("unable to extract port for the gRPC endpoint: %w", err)
+		}
+	}
+
+	if cfg.HTTP != nil {
+		if _, err := extractPortFromEndpoint(cfg.HTTP.Endpoint); err != nil {
+			return fmt.Errorf("unable to extract port for the Thrift HTTP endpoint: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Unmarshal a config.Parser into the config struct.
+func (cfg *Config) Unmarshal(componentParser *config.Map) error {
+	if componentParser == nil || len(componentParser.AllKeys()) == 0 {
+		return fmt.Errorf("empty config for Jaeger receiver")
+	}
+
+	// UnmarshalExact will not set struct properties to nil even if no key is provided,
+	// so set the protocol structs to nil where the keys were omitted.
+	err := componentParser.UnmarshalExact(cfg)
+	if err != nil {
+		return err
+	}
+
+	protocols, err := componentParser.Sub(protocolsFieldName)
+	if err != nil {
+		return err
+	}
+
+	if !protocols.IsSet(protoGRPC) {
+		cfg.GRPC = nil
```

Comment:
Not sure, why is this required?

Reply:
just a redundant check.

Reply:
I think it's more than just a check. The way it works is that if `grpc` is not included on the protocol list, the receiver will disable it.

This is not consistent with how `http` is used. It will be always started (if not provided specific config, then using the defaults). 

If this was intended to map how e.g. Jaeger receiver works, I think similar check and setting nil is required for HTTP as well.

Reply:
yes, I referred to the implementation of jaeger receiver. I will update it.